source/menu/menu_coop.qc

```diff
@@ -0,0 +1,199 @@
+
+float player_name_cursor;
+string player_name;
+
+float server_ip_cursor;
+string server_ip;
+
+float server_hostname_cursor;
+string server_hostname;
+
+float server_password_cursor;
+string server_password;
+
+float server_port_cursor;
+string server_port;
+
+int num_cached_servers;
+
+void() Menu_Coop_Name_Update =
+{
+	static float initialized;
+	if (!initialized)
+	{
+		player_name = cvar_string("name");
+		player_name_cursor = strlen(player_name);
+		initialized = true;
+	}
+	else
+	{
+		cvar_set("name", player_name);
+	}
+};
+
+void() Menu_Coop_Hostname_Update =
+{
+	static float initialized;
+	if (!initialized)
+	{
+		server_hostname = cvar_string("hostname");
+		server_hostname_cursor = strlen(server_hostname);
+		initialized = true;
+	}
+	else
+	{
+		cvar_set("hostname", server_hostname);
+	}
+};
+
+void() Menu_Coop_Port_Update =
+{
+	static float initialized;
+	if (!initialized)
+	{
+		server_port = cvar_string("sv_port");
+		server_port_cursor = strlen(server_port);
+		initialized = true;
+	}
+	else
+	{
+		cvar_set("sv_port", server_port);
+	}
+};
+
+void() Menu_Coop_Password_Update =
+{
+	static float initialized;
+	if (!initialized)
+	{
+		server_password = cvar_string("password");
+		server_password_cursor = strlen(server_password);
+		initialized = true;
+	}
+	else
+	{
+		cvar_set("password", server_password);
+	}
+};
+
+void() Menu_Coop_Browse_Refresh =
+{
+	static float refreshtime;
+
+	if (refreshtime > time)
+		return;
+
+	refreshtime = time + 0.5;
+
+	resethostcachemasks();
+	// TODO: figure out why this returns 0 servers
+	// sethostcachemaskstring(0, gethostcacheindexforkey("gamedir"), cvar_string("game"), SLIST_TEST_EQUAL);
+	sethostcachesort(gethostcacheindexforkey("ping"), false);
+	refreshhostcache(false);
+	resorthostcache();
+
+	num_cached_servers = (int)gethostcachevalue(SLIST_HOSTCACHEVIEWCOUNT);
+};
+
+void() Menu_Coop_Browse =
+{
+
+	Menu_Coop_Browse_Refresh();
+
+	Menu_DrawBackground();
+
+	Menu_DrawTitle("BROWSE SERVERS");
+
+	// draw server list
+	sui_set_align([SUI_ALIGN_CENTER, SUI_ALIGN_CENTER]);
+	static vector serverlist_scrollofs;
+	vector vsize = sui_current_frame_size();
+	Menu_ServerList("cbm_serverlist", [0, 0], [vsize.x, vsize.y / 2], serverlist_scrollofs, num_cached_servers);
+
+	Menu_Button(-1, "cbm_back", "BACK", "Return to Join Game Menu.") ? current_menu = MENU_COOPJOIN : 0;
+
+	sui_pop_frame();
+};
+
+void() Menu_Coop_Direct =
+{
+	Menu_Coop_Password_Update();
+
+	Menu_DrawBackground();
+
+	Menu_DrawTitle("DIRECT IP/ID");
+
+	Menu_Button(1, "cdm_serverip", "SERVER IP/ROOM ID", "IP or ID for the Server (typically starts with /).");
+	Menu_TextInput("cdm_serverip", 1, server_ip, server_ip_cursor);
+
+	Menu_Button(2, "cdm_serverpassword", "SERVER PASSWORD", "Password for the Match set by the Host.");
```

Comment:
We should specify this is optional, there have been usercomplaints about this in the past.

Reply:
Resolved in e02d043a66b338eeaff1493dcb5cb3b59d627289